### PR TITLE
[Ready for Review] Refining CI tests: Run test on changed files in PR

### PR DESCRIPTION
**Pull request by sgibson91**
### Summary

When our test scripts are executed in a Pull Request, they check _all_ files in the repo rather than selectively the ones that have changed. This means that sometimes, to get tests to pass and allow merging, we have to make changes to someone else's work. This makes it much more difficult to think of a PR as a single piece of work.

In this PR, I have updated our test scripts (`lorem-ipsums.py` and `no-bad-latin.py`, `test-file-size.sh` is unchanged) to take the Pull Request number as an argument and check only the files that have been added or modified in the PR. Therefore, a contributor doesn't have to worry about fixing other people's work _at the same time_ as contributing their own. A call is made to the GitHub API to return a list of changed files in `pull_files.py`.

The GitHub Actions that have run the tests in CI have been reconfigured such that they run in "changed files" mode for PRs, but then _all_ files are checked on pushes to master.

**Remain TODO:** Update the required statuses in the branch protection rules as I've renamed things.

### List of changes proposed in this PR (pull-request)

* `lorem-ipsums.py`
* `no-bad-latin.py`
* `lorem-ipsums.yml`
* `no-bad-latin.yml`
* `pull_files.py`
* Update status badges in README

### What should a reviewer concentrate their feedback on?

- [ ] Is this a sensible way to run CI?
- [ ] Everything looks ok?

### Acknowledging contributors

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [x] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file:  Please add @martinagvilas for tests

### CODE REVIEW

**Comment by malvikasharan** [comment]
Thank you @sgibson91 🙌 LGTM.

**Comment by sgibson91** [comment]
I'm actually no longer convinced this will work how I think it should. Holding off merging for now.

**Comment by sgibson91** [comment]
### Update for @martinagvilas

The problem in this PR is that the steps with the conditionals `if: github.event.push` don't run when the PR is merged to master, e.g. this line: https://github.com/alan-turing-institute/the-turing-way/blob/c4f47e64b726455542789b081650c32ff81845a4/.github/workflows/lorem-ipsums.yml#L45

Also, the actual test scripts themselves have been changed to pull the changed files by making HTTP requests to the GitHub API, e.g.: https://github.com/alan-turing-institute/the-turing-way/blob/c4f47e64b726455542789b081650c32ff81845a4/tests/lorem-ipsums.py#L40-L56

I had tried using [this GitHub Action](https://github.com/marketplace/actions/get-changed-files) but it is buggy (Issue: https://github.com/lots0logs/gh-action-get-changed-files/issues/8).

Hopefully this is helpful! Let me know if you'd like to co-work on this at any time! 🚀

**Comment by martinagvilas** [comment]
Hey @sgibson91, thanks for filling me in 🚀 I'm interested in co-working on this to make it work!

> The problem in this PR is that the steps with the conditionals `if: github.event.push` don't run when the PR is merged to master, e.g. this line:

Did you try..?
`if: github.event_name == 'push' && github.ref == 'refs/heads/master'` instead of  `if: github.event.push`

and is  `if: github.event.pull_request` working as expected?

Also: 

```
pr_url = f"https://api.github.com/repos/alan-turing-institute/the-turing-way/pulls/{pr_num}/files" 
resp = requests.get(pr_url) 
```
 ... looks super cool! Did you manage to get a json of the changed files with that?

**Comment by sgibson91** [comment]
Yay! Let's try and connect during Malvika's co-working hours or we can schedule something else if that doesn't work for you :)

> Did you try..?
> `if: github.event_name == 'push' && github.ref == 'refs/heads/master'` instead of `if: github.event.push`

I did not! That's the first thing on the to-do list!

> and is `if: github.event.pull_request` working as expected?

I believe so but it's been a while since I tinkered with this, so I should probably convince myself of this again.

> ... looks super cool! Did you manage to get a json of the changed files with that?

Yes! The below lines it looping through the JSON response which is a list of dictionaries relating to the changed files in the PR. I'm just looping over those dictionaries to append the filenames to a list to be checked individually. I freaking LOVE the GitHub API, it's so cool!

**Comment by martinagvilas** [comment]
> Yay! Let's try and connect during Malvika's co-working hours or we can schedule something else if that doesn't work for you :)

Great 🚀  I will check what day works for me and let you know!

> Yes! The below lines it looping through the JSON response which is a list of dictionaries relating to the changed files in the PR. I'm just looping over those dictionaries to append the filenames to a list to be checked individually. I freaking LOVE the GitHub API, it's so cool!

🤯 🤯 I spent two hours the other day trying to do this without the API 😢 But so glad to know now that it exists!

**Comment by martinagvilas** [comment]
Hey @sgibson91 are you going to be around during the collaboration café tomorrow? So we can chat about this. If not any other day of the week works for me, but couldn't find the co-working hackmd of this week to sign up!

**Comment by sgibson91** [comment]
Hi @martinagvilas, I'm not sure about tomorrow. There's a group of us giving talks at a conference next week and we have earmarked tomorrow afternoon to practise them with Kirstie (we weren't given a lot of prep time by the organisers!). The co-working hackmd is [here](https://hackmd.io/@malvikasharan/TW-coworking) but I suspect Malvika has been too busy to update it (she is also giving a talk!). I could send you a Zoom link to co-work on Thursday if that would be useful?

**Comment by martinagvilas** [comment]
@sgibson91 the zoom link for Thursday is fine by me, but if you are too busy this week we can easily leave it for the next one! Either way is ok for me

**Comment by sgibson91** [comment]
I will let you know, thanks!

**Comment by sgibson91** [comment]
@martinagvilas I double checked my calendar and can make the Cafe today!

**Comment by martinagvilas** [comment]
Great! I'll see you there

**Comment by sgibson91** [comment]
I believe this is now ready to merge. I have requested @tallamjr's review as well since he has much GitHub and Python fu 🤗

**Comment by sgibson91** [comment]
@all-contributors please add @martinagvilas for tests

File: tests/lorem-ipsums.py [hunk 1]
```diff
@@ -21,25 +37,67 @@ def remove_comments(text_string):
...
+        text = remove_comments(text)
+        if bad_phrase in text.lower():
+            failed.append(filename.name)
+
+    return failed
+
+
+def check_all_files(
+    bad_phrase=BAD_PHRASE,
+    directory_to_check=os.path.join(ABSOLUTE_HERE, "book", "content"),
```
Comment (martinagvilas) [inline tests/lorem-ipsums.py:69]: I think this path might not be found and trigger an error with the new build of the book

File: tests/no-bad-latin.py [hunk 2]
```diff
@@ -89,29 +104,35 @@ def read_and_check_files(files):
     return failing_files
 
 
-def construct_error_message(files_dict):
-    """Function to construct an error message pointing out where bad latin
-	phrases appear in lines of text
+def get_all_files(directory=os.path.join(ABSOLUTE_HERE,"book", "content")):
```
Comment (martinagvilas) [inline tests/no-bad-latin.py:107]: If this function is called without an argument the content path might trigger an error in this case too.

File: tests/lorem-ipsums.py (hunk 1 above)
Comment (sgibson91) [reply tests/lorem-ipsums.py]: Good catch on these, I've updated both!

File: .gitignore [hunk 3]
```diff
@@ -10,3 +10,5 @@ book/website/Pipfile\.lock
 
 workshops/boost-research-reproducibility-binder/workshop-presentations/*pptx
 .Rproj.user
+
+*.pyc
```
Comment (sgibson91) [inline .gitignore:14]:
```suggestion
**/__pycache__/
```
Git-ignore the pycache folder rather than individual compiled files

**Pull request merged**